news/backends/sfmc.py

```diff
@@ -10,19 +12,95 @@
 from news.backends.common import NewsletterException, NewsletterNoResultsException
 
 
+HERD_TIMEOUT = 60
+AUTH_BUFFER = 300  # 5 min
+MAX_BUFFER = HERD_TIMEOUT + AUTH_BUFFER
+
+
 class ETRefreshClient(ET_Client):
+    token_cache_key = 'backends:sfmc:auth:tokens'
+    authTokenExpiresIn = None
+    token_property_names = [
+        'authToken',
+        'authTokenExpiration',
+        'internalAuthToken',
+        'refreshKey',
+    ]
+    _old_authToken = None
+
+    def token_is_expired(self):
+        """Report token is expired between 5 and 6 minutes early
+
+        Having the expiration be random helps prevent multiple basket
+        instances simultaneously requesting a new token from SFMC,
+        a.k.a. the Thundering Herd problem.
+        """
+        if self.authTokenExpiration is None:
+            return True
+
+        time_buffer = randint(1, HERD_TIMEOUT) + AUTH_BUFFER
+        return time() + time_buffer > self.authTokenExpiration
+
+    def refresh_auth_tokens_from_cache(self):
+        """Refresh the auth token and other values from cache"""
+        if self.authToken is not None and time() + MAX_BUFFER < self.authTokenExpiration:
+            # no need to refresh if the current tokens are still good
+            return
+
+        tokens = cache.get(self.token_cache_key)
+        if tokens:
+            for prop, value in tokens.items():
+                setattr(self, prop, value)
```

Comment:
not necessarily hostile, but we're crossing a network here and bit flips and other data corruption are possible